### PR TITLE
Expose supported versions

### PR DESCRIPTION
**Pull request by provokateurin**
Closes https://github.com/provokateurin/nextcloud-neon/issues/322

Already making use of new dart features :)

@adil192 does this interface work for you?

### CODE REVIEW

**Comment by adil192** [comment]
This is perfect

So far the only Dart 3 feature I'm using is not needing `break` statements in switch cases lol

File: packages/nextcloud/lib/src/version_supported.dart [hunk 1]
```diff
@@ -1,33 +1,68 @@
...
+        (capabilities ?? (await getCapabilities()).ocs.data).version.major == coreSupportedVersion.version,
+        coreSupportedVersion,
+      );
 }
 
 // ignore: public_member_api_docs
 extension NewsVersionSupported on NextcloudNewsClient {
   /// Checks if the app on the server is supported by the client
-  Future<bool> isSupported() async {
+  Future<(bool, _SupportedVersion<String>)> isSupported() async {
```
Comment (Leptopoda) [inline packages/nextcloud/lib/src/version_supported.dart:44]: Somehow I dislike this interface.

Tuples (or however dart calls them) got introduced so we don't need to create such dummy objects and we still have one here.

Also doesn't the caller already know the app they are testing for?
I'd suggest `(isSupported, version)` or if we really need the app name `(isSupported, version, appName)`

File: packages/neon/neon/lib/src/pages/home.dart [hunk 2]
```diff
@@ -60,30 +60,33 @@ class _HomePageState extends State<HomePage> {
             ...appsResult.data!.map((final a) => a.id),
           ]) {
             try {
-              bool? supported;
+              late bool supported;
```
Comment (Leptopoda) [inline packages/neon/neon/lib/src/pages/home.dart:63]: dart3 could 
```
token = switch (charCode) {
  slash || star || plus || minus => operator(charCode),
  comma || semicolon => punctuation(charCode),
  >= digit0 && <= digit9 => number(),
  _ => throw invalid()
};
```

**Review by Leptopoda — CHANGES_REQUESTED** [review]
at least the switch could be cleaned up
Also just reviewed online and didn't test locally yet :)

File: packages/neon/neon/lib/src/pages/home.dart (hunk 2 above)
Comment (provokateurin) [reply packages/neon/neon/lib/src/pages/home.dart:63]: Ah yes I didn't think about that :)

File: packages/nextcloud/lib/src/version_supported.dart (hunk 1 above)
Comment (provokateurin) [reply packages/nextcloud/lib/src/version_supported.dart:44]: Hm yeah you're right, the app doesn't really add any info

File: packages/nextcloud/lib/src/version_supported.dart [hunk 3]
```diff
@@ -1,33 +1,57 @@
...
+  newsSupportedVersion,
+  notesSupportedVersion,
+];
+
 // ignore: public_member_api_docs
 extension CoreVersionSupported on NextcloudCoreClient {
   /// Checks if the app on the server is supported by the client
-  Future<bool> isSupported([final NextcloudCoreServerCapabilities_Ocs_Data? capabilities]) async =>
-      (capabilities ?? (await getCapabilities()).ocs.data).version.major == 26;
+  Future<(bool, int)> isSupported([
```
Comment (provokateurin) [inline packages/nextcloud/lib/src/version_supported.dart:22]: The version that is returned is always just the constant and could be queried separately, but I think this interface makes it a little easier to work with.

File: packages/nextcloud/lib/src/version_supported.dart [hunk 4]
```diff
@@ -1,33 +1,57 @@
...
+/// All supported versions
+const supportedVersions = [
+  coreSupportedVersion,
+  newsSupportedVersion,
+  notesSupportedVersion,
+];
+
 // ignore: public_member_api_docs
 extension CoreVersionSupported on NextcloudCoreClient {
   /// Checks if the app on the server is supported by the client
```
Comment (Leptopoda) [inline packages/nextcloud/lib/src/version_supported.dart:21]: I know I'm getting nit picky here but the doc comments also need adjustments

File: packages/nextcloud/lib/src/version_supported.dart (hunk 3 above)
Comment (Leptopoda) [reply packages/nextcloud/lib/src/version_supported.dart:22]: Yeah but we don't hide it so if somebody wants to look it up separately they can.

File: packages/nextcloud/lib/src/version_supported.dart (hunk 4 above)
Comment (provokateurin) [reply packages/nextcloud/lib/src/version_supported.dart:21]: What do you want to have here instead?

File: packages/nextcloud/lib/src/version_supported.dart [hunk 5]
```diff
@@ -1,33 +1,57 @@
...
+const coreSupportedVersion = 26;
+
+/// API version of the news app supported
+const newsSupportedVersion = 'v1-3';
+
+/// API version of the notes app supported
+const notesSupportedVersion = 1;
+
+/// All supported versions
+const supportedVersions = [
```
Comment (Leptopoda) [inline packages/nextcloud/lib/src/version_supported.dart:13]: Where do we need this?

I don't think a list is that helpful as here the app name is relevant. `Maybe a Map<appName, versionSupported>` ?

Comment (provokateurin) [reply packages/nextcloud/lib/src/version_supported.dart:13]: I noticed it too, removed it.

File: packages/nextcloud/lib/src/version_supported.dart (hunk 4 above)
Comment (Leptopoda) [reply packages/nextcloud/lib/src/version_supported.dart:14]: I'd mention that it what it returns.
How should I know what the second parameter represents when just using code highlighting in an IDE.
 
 sadly I couldn't find anything in the effective-dart:documentation guide regarding this issue.
 ```dart
 /// returns something cool
(bool issupported, int minVersion) checkSomething() {

}
 ```
 
 Also the IDE does not really highlight the names I gave it in the above example o the best solution might be adding a 'Returns the result of the check and the minimum supported version.'
 
 Or maybe a code example :thinking:

**Review by Leptopoda — APPROVED** [review]

**Pull request merged**